subsys/mgmt/hawkbit/hawkbit.c

```diff
@@ -0,0 +1,1267 @@
+	if (ret < 0) {
+		LOG_ERR("Unable to parse deployment base");
+		goto cleanup;
+	}
+
+	nvs_read(&fs, ADDRESS_ID, &action_id, sizeof(action_id));
+
+	if (action_id == (int32_t)hb_context.json_action_id) {
+		LOG_INF("Preventing repeated attempt to install %d",
+			hb_context.json_action_id);
+		hb_context.dl.http_content_size = 0;
+		memset(hb_context.url_buffer, 0, sizeof(hb_context.url_buffer));
+		hb_context.url_buffer_size = URL_BUFFER_SIZE;
+		snprintk(hb_context.url_buffer, hb_context.url_buffer_size,
+			 "%s/%s-%s/deploymentBase/%d/feedback",
+			 HAWKBIT_JSON_URL, CONFIG_BOARD, device_id,
+			 hb_context.json_action_id);
+
+		if (!send_request(HTTP_POST, HAWKBIT_REPORT,
+				  HAWKBIT_STATUS_FINISHED_SUCCESS,
+				  HAWKBIT_STATUS_EXEC_CLOSED)) {
+			LOG_ERR("Error when querying from hawkbit");
+			hb_context.code_status = HAWKBIT_NETWORKING_ERROR;
+			goto cleanup;
+		}
+
+		hb_context.code_status = HAWKBIT_OK;
+		goto cleanup;
+	}
+
+	LOG_INF("Ready to install update");
+
+	hb_context.dl.http_content_size = 0;
+	memset(hb_context.url_buffer, 0, sizeof(hb_context.url_buffer));
+	hb_context.url_buffer_size = URL_BUFFER_SIZE;
+
+	snprintk(hb_context.url_buffer, hb_context.url_buffer_size, "%s",
+		 download_http);
+
+	flash_img_init(&hb_context.flash_ctx);
+
+	if (!send_request(HTTP_GET, HAWKBIT_DOWNLOAD,
+			  HAWKBIT_STATUS_FINISHED_NONE,
+			  HAWKBIT_STATUS_EXEC_NONE)) {
+		LOG_ERR("Send request failed");
+		hb_context.code_status = HAWKBIT_NETWORKING_ERROR;
+		goto cleanup;
+	}
+
+	if (hb_context.code_status == HAWKBIT_DOWNLOAD_ERROR) {
+		goto cleanup;
+	}
+
+	if (boot_request_upgrade(BOOT_UPGRADE_TEST)) {
+		LOG_ERR("Download failed");
+		hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
+	} else {
+		hb_context.code_status = HAWKBIT_UPDATE_INSTALLED;
+		hawkbit_device_acid_update(hb_context.json_action_id);
+	}
+
+	hb_context.dl.http_content_size = 0;
+
+cleanup:
+	cleanup_connection();
+
+error:
+	free(hb_context.response_data);
+	return hb_context.code_status;
+}
+
+static void autohandler(struct k_work *work)
+{
+	switch (hawkbit_probe()) {
+	case HAWKBIT_UNCONFIRMED_IMAGE:
+		LOG_ERR("Image is unconfirmed");
+		LOG_ERR("Rebooting to previous confirmed image.");
+		sys_reboot(SYS_REBOOT_WARM);
+		break;
+
+	case HAWKBIT_NO_UPDATE:
+		LOG_INF("No update found");
+		break;
+
+	case HAWKBIT_CANCEL_UPDATE:
+		LOG_INF("Hawkbit update Cancelled from server");
+		break;
+
+	case HAWKBIT_OK:
+		LOG_INF("Image is already updated");
+		break;
+
+	case HAWKBIT_UPDATE_INSTALLED:
+		LOG_INF("Update Installed. Please Reboot");
+		break;
+
+	case HAWKBIT_DOWNLOAD_ERROR:
+		LOG_INF("Update Failed");
+		break;
+
+	case HAWKBIT_NETWORKING_ERROR:
+		LOG_INF("Network Error");
+		break;
+
+	case HAWKBIT_METADATA_ERROR:
+		LOG_INF("Metadata error");
+		break;
+	}
+
+	k_delayed_work_submit(&hawkbit_work_handle, K_MSEC(poll_sleep));
+}
+
+void hawkbit_autohandler(void)
+{
+	k_delayed_work_init(&hawkbit_work_handle, autohandler);
```

Comment:
Maybe an improvement for the next iteration, but you should really consider adding a separate thread/workqueue in this library for the automode (perhaps automode shall be optionally enabled with Kconfig, to save resources in case it's not used). Putting such a lenghty function calls like `hawkbit_probe()` into a system workqueue is not a good idea (as far as I understand this function might even wait for an entire firmare to downlad).

 It might work in a simple sample, but in more advanced apps, where the FOTA library is just an addition, it's not acceptable to block the system workqueue for so long.

Reply:
Thanks for the improvement feedback. Noted for the next iteration to solve the NVS and implement separate thread/workqueue to handle the FOTA library.